File: backend/src/services/ai_settings_service.py
"""AI Settings service for business logic."""
from sqlalchemy.orm import Session
from typing import Optional

try:
    from repositories.ai_settings_repository import AISettingsRepository
    from schemas.ai_settings import AISettingsCreate, AISettingsUpdate, AISettingsResponse
    from models.ai_settings import AISettings
except ImportError:
    from src.repositories.ai_settings_repository import AISettingsRepository
    from src.schemas.ai_settings import AISettingsCreate, AISettingsUpdate, AISettingsResponse
    from src.models.ai_settings import AISettings


class AISettingsService:
    """Service for AI Settings business logic."""
    
    def __init__(self, db: Session):
        self.repository = AISettingsRepository(db)
        self.db = db
    
    def get_settings(self) -> Optional[AISettingsResponse]:
        """获取AI设置"""
        settings = self.repository.get_or_create_default()
        if settings:
            return AISettingsResponse.model_validate(settings)
        return None
    
    def create_settings(self, settings_data: AISettingsCreate) -> AISettingsResponse:
        """创建AI设置"""
        # 验证区域点数量
        if settings_data.danger_zone and len(settings_data.danger_zone) != 4:
            raise ValueError("危险区域必须包含4个点")
        if settings_data.warning_zone and len(settings_data.warning_zone) != 4:
            raise ValueError("警告区域必须包含4个点")
        
        # 如果提供了camera_id，验证摄像头是否存在
        if settings_data.camera_id:
            from src.repositories.camera_repository import CameraRepository
            camera_repo = CameraRepository(self.db)
            camera = camera_repo.get_by_id(settings_data.camera_id)
            if not camera:
                raise ValueError(f"摄像头 {settings_data.camera_id} 不存在")
            # 自动填充摄像头信息
            settings_data.camera_name = camera.name
            settings_data.camera_url = camera.url
        
        settings = self.repository.create(settings_data)
        return AISettingsResponse.model_validate(settings)
    
    def update_settings(self, settings_id: int, settings_data: AISettingsUpdate) -> Optional[AISettingsResponse]:
        """更新AI设置"""
        # 验证区域点数量
        if settings_data.danger_zone is not None and len(settings_data.danger_zone) != 4:
            raise ValueError("危险区域必须包含4个点")
        if settings_data.warning_zone is not None and len(settings_data.warning_zone) != 4:
            raise ValueError("警告区域必须包含4个点")
        
        # 如果更新了camera_id，验证摄像头是否存在并更新相关信息
        if settings_data.camera_id:
            from src.repositories.camera_repository import CameraRepository
            camera_repo = CameraRepository(self.db)
            camera = camera_repo.get_by_id(settings_data.camera_id)
            if not camera:
                raise ValueError(f"摄像头 {settings_data.camera_id} 不存在")
            # 自动更新摄像头信息
            settings_data.camera_name = camera.name
            settings_data.camera_url = camera.url
        
        settings = self.repository.update(settings_id, settings_data)
        if settings:
            return AISettingsResponse.model_validate(settings)
        return None
    
    def delete_settings(self, settings_id: int) -> bool:
        """删除AI设置"""
        return self.repository.delete(settings_id)
    
    def bind_camera(self, settings_id: int, camera_id: str) -> Optional[AISettingsResponse]:
        """绑定摄像头到AI设置"""
        from src.repositories.camera_repository import CameraRepository
        camera_repo = CameraRepository(self.db)
        camera = camera_repo.get_by_id(camera_id)
        
        if not camera:
            raise ValueError(f"摄像头 {camera_id} 不存在")
        
        update_data = AISettingsUpdate(
            camera_id=camera_id,
            camera_name=camera.name,
            camera_url=camera.url
        )
        
        return self.update_settings(settings_id, update_data)
    
    def unbind_camera(self, settings_id: int) -> Optional[AISettingsResponse]:
        """解绑摄像头"""
        update_data = AISettingsUpdate(
            camera_id=None,
            camera_name=None,
            camera_url=None
        )
        
        return self.update_settings(settings_id, update_data)

File: src/services/aiSettingsService.ts
/**
 * AI Settings Service - Frontend API Client
 * Handles communication with backend AI settings API
 */

/**
 * 坐标点接口
 */
export interface Point {
  x: number; // 0-1
  y: number; // 0-1
}

/**
 * AI 设置接口
 */
export interface AISettings {
  id: number;
  enabled: boolean;
  camera_id: string | null;
  camera_name: string | null;
  camera_url: string | null;
  confidence_threshold: number; // 0-100
  danger_zone: Point[] | null; // 4个点
  warning_zone: Point[] | null; // 4个点
  sound_alarm: boolean;
  visual_alarm: boolean;
  auto_screenshot: boolean;
  alarm_cooldown: number; // 秒
  created_at: string;
  updated_at: string;
}

/**
 * 创建 AI 设置数据
 */
export interface AISettingsCreate {
  enabled?: boolean;
  camera_id?: string | null;
  confidence_threshold?: number;
  danger_zone?: Point[] | null;
  warning_zone?: Point[] | null;
  sound_alarm?: boolean;
  visual_alarm?: boolean;
  auto_screenshot?: boolean;
  alarm_cooldown?: number;
}

/**
 * 更新 AI 设置数据
 */
export interface AISettingsUpdate {
  enabled?: boolean;
  camera_id?: string | null;
  confidence_threshold?: number;
  danger_zone?: Point[] | null;
  warning_zone?: Point[] | null;
  sound_alarm?: boolean;
  visual_alarm?: boolean;
  auto_screenshot?: boolean;
  alarm_cooldown?: number;
}

/**
 * AI Settings Service class
 */
export class AISettingsService {
  private baseUrl: string;

  constructor(baseUrl: string = 'http://127.0.0.1:8000') {
    this.baseUrl = baseUrl;
  }

  /**
   * 获取 AI 设置
   */
  async getSettings(): Promise<AISettings> {
    const response = await fetch(`${this.baseUrl}/api/ai-settings`);
    if (!response.ok) {
      throw new Error(`Failed to fetch AI settings: ${response.statusText}`);
    }
    return await response.json();
  }

  /**
   * 创建 AI 设置
   */
  async createSettings(data: AISettingsCreate): Promise<AISettings> {
    const response = await fetch(`${this.baseUrl}/api/ai-settings`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(data),
    });
    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to create AI settings');
    }
    return await response.json();
  }

  /**
   * 更新 AI 设置
   */
  async updateSettings(settingsId: number, data: AISettingsUpdate): Promise<AISettings> {
    const response = await fetch(`${this.baseUrl}/api/ai-settings/${settingsId}`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(data),
    });
    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to update AI settings');
    }
    return await response.json();
  }

  /**
   * 删除 AI 设置
   */
  async deleteSettings(settingsId: number): Promise<void> {
    const response = await fetch(`${this.baseUrl}/api/ai-settings/${settingsId}`, {
      method: 'DELETE',
    });
    if (!response.ok) {
      throw new Error('Failed to delete AI settings');
    }
  }

  /**
   * 绑定摄像头
   */
  async bindCamera(settingsId: number, cameraId: string): Promise<AISettings> {
    const response = await fetch(
      `${this.baseUrl}/api/ai-settings/${settingsId}/bind-camera/${cameraId}`,
      {
        method: 'POST',
      }
    );
    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to bind camera');
    }
    return await response.json();
  }

  /**
   * 解绑摄像头
   */
  async unbindCamera(settingsId: number): Promise<AISettings> {
    const response = await fetch(
      `${this.baseUrl}/api/ai-settings/${settingsId}/unbind-camera`,
      {
        method: 'POST',
      }
    );
    if (!response.ok) {
      throw new Error('Failed to unbind camera');
    }
    return await response.json();
  }
}

/**
 * Singleton instance
 */
export const aiSettingsService = new AISettingsService();
